src/state_oorset_ps.erl

```diff
@@ -0,0 +1,890 @@
+%%
+%% Copyright (c) 2015-2016 Christopher Meiklejohn.  All Rights Reserved.
+%% Copyright (c) 2007-2012 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Optimised ORSet CRDT with provenance semirings:
+%%      observed-remove set without tombstones
+
+-module(state_oorset_ps).
+-author("Junghun Yoo <junghun.yoo@cs.ox.ac.uk>").
+
+-behaviour(type).
+-behaviour(state_type).
+
+-define(TYPE, ?MODULE).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([new/0, new/1]).
+-export([mutate/3, delta_mutate/3, merge/2]).
+-export([query/1, equal/2, is_bottom/1, is_inflation/2, is_strict_inflation/2]).
+-export([join_decomposition/1]).
+
+-export([subtract_all_events/2,
+         add_elem_with_dot/3,
+         get_events_from_provenance/1,
+         cross_provenance/2,
+         subtract_removed/2]).
+
+-export_type([state_oorset_ps/0,
+              delta_state_oorset_ps/0,
+              state_oorset_ps_op/0]).
+
+-opaque state_oorset_ps() :: {?TYPE, payload()}.
+-opaque delta_state_oorset_ps() :: {?TYPE, {delta, payload()}}.
+-type delta_or_state() :: state_oorset_ps() | delta_state_oorset_ps().
+-type payload() :: {ps_data_store(), ps_filtered_out_events(), ps_all_events()}.
+-type element() :: term().
+-type state_oorset_ps_op() :: {add, element()}
+                               | {add_all, [element()]}
+                               | {rmv, element()}
+                               | {rmv_all, [element()]}.
+
+%% Provenanve semiring related.
+-export_type([ps_event_id/0]).
+
+-type ps_object_id() :: binary().
+-type ps_replica_id() :: term().
+-type ps_event_id() :: {ps_object_id(), ps_replica_id()}.
+-type ps_event() :: {ps_event_id(), pos_integer()}.
+-type ps_dot() :: ordsets:ordsets(ps_event()).
```

Comment:
A dot is a single reason to be existing. Normally, it contains a single event (after the 'add' operation), but it can have multiple events (after the product(), a dot contains both left & right operands' events). I am adding documentation now, so it will be added.